pg_analytics/src/hooks/udf.rs

```diff
@@ -0,0 +1,223 @@
+                    num_rows = arg_arr.len();
+                }
+            }
+        }
+
+        // Get function name - will always be a scalar value
+        if let ColumnarValue::Scalar(ScalarValue::Utf8(Some(funcname))) = &args[0] {
+            // Turn all arguments into arrays of result length
+            let mut arg_arrays = vec![];
+            for arg in args.iter().take(num_args).skip(1) {
+                arg_arrays.push(arg.clone().into_array(num_rows)?);
+            }
+
+            // Call function!
+            // Follows the internal logic of FunctionCall9Coll
+            let func_oid = func_oid_from_signature(funcname, arg_oids)?;
+
+            // Create function call struct
+            let flinfo = PgBox::<pg_sys::FmgrInfo>::alloc0();
+            pg_sys::fmgr_info(func_oid, flinfo.as_ptr());
+            // Prefer to use std::mem::offset_of! but it is still unstable as of rust 1.76.0.
+            let fcinfo_size = max(
+                std::mem::size_of::<pg_sys::FunctionCallInfoBaseData>(),
+                memoffset::offset_of!(pg_sys::FunctionCallInfoBaseData, args)
+                    + std::mem::size_of::<pg_sys::NullableDatum>() * (num_args - 1),
+            );
+            let fcinfo = pg_sys::palloc0(fcinfo_size) as *mut pg_sys::FunctionCallInfoBaseData;
+            (*fcinfo).flinfo = flinfo.as_ptr();
+            (*fcinfo).context = std::ptr::null_mut();
+            (*fcinfo).resultinfo = std::ptr::null_mut();
+            (*fcinfo).fncollation = pg_sys::Oid::INVALID;
+            (*fcinfo).isnull = false;
+            (*fcinfo).nargs = num_args as i16 - 1;
+
+            // Call function on each set of arguments and push to result vector
+            let mut result_vec = vec![];
+            for row_index in 0..num_rows {
+                for arg_index in 1..num_args {
+                    let fcinfo_arg = (*fcinfo).args.as_mut_ptr().add(arg_index - 1);
+                    (*fcinfo_arg).value = arg_arrays[arg_index - 1]
+                        .get_datum(row_index)?
+                        .ok_or(DataFusionError::Internal("No datum for type".to_string()))?;
+                    (*fcinfo_arg).isnull = false;
+                }
+
+                let result_datum = (*(*fcinfo).flinfo)
+                    .fn_addr
+                    .ok_or(DataFusionError::Internal(
+                        "Invalid function address for udf".to_string(),
+                    ))?(fcinfo);
+                result_vec.push((result_datum, false));
+            }
+
+            // UDF return type
+            let rettype = pg_sys::get_func_rettype(func_oid);
+
+            Ok(ColumnarValue::Array(
+                result_vec
+                    .into_iter()
+                    .into_arrow_array(rettype.into(), PgTypeMod(-1))?,
+            ))
+        } else {
+            Err(DataFusionError::Internal("No funcname".to_string()))
+        }
+    });
+
+    drop(memory_context);
```

Comment:
Is this drop necessary? It should go out of scope at the end of the block. I know memory contexts are weird, so if you really do need this please just leave a comment.

Reply:
Also, instead of creating a new memory context, can we just use `CurrentMemoryContext`?

Reply:
It's better to create a separate memory context for this use case to keep the memory separate. You're right about the `drop`, though - that's a remnant from a previous implementation I was working on.